tests/unit/test_dimension_utilities.py

```diff
@@ -304,6 +311,141 @@ def test_prefetch_dimension_variables(self, mock_get_opendap_nc4):
+            'tests/data/ATL16_prefetch_bnds.dmr'
+        )
+
+        with self.subTest('Variable has cell alignment and bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable('/variable_edge_has_bnds')))
+
+        with self.subTest('Variable has no cell alignment and has bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable('/variable_no_edge_has_bnds')))
+
+        with self.subTest('Variable has cell alignment and no bounds'):
+            self.assertTrue(needs_bounds(varinfo_bounds.get_variable('/variable_edge_no_bnds')))
+
+        with self.subTest('Variable has no cell alignment and no bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable('/variable_no_edge_no_bnds')))
+
+
+    def test_get_bounds_array(self):
+        """ Ensure that the expected bounds array is created given
+            the input dimension variable values.
+
+        """
+        prefetch_dataset = Dataset('tests/data/ATL16_prefetch.nc4', 'r')
+        dimension_path = '/npolar_grid_lat'
+
+        expected_bounds_array = np.array([[90.0, 89.0], [89.0, 88.0], [88.0, 87.0], [87.0, 86.0],
+                                 [86.0, 85.0], [85.0, 84.0], [84.0, 83.0], [83.0, 82.0],
+                                 [82.0, 81.0], [81.0, 80.0], [80.0, 79.0], [79.0, 78.0],
+                                 [78.0, 77.0], [77.0, 76.0], [76.0, 75.0], [75.0, 74.0],
+                                 [74.0, 73.0], [73.0, 72.0], [72.0, 71.0], [71.0, 70.0],
+                                 [70.0, 69.0], [69.0, 68.0], [68.0, 67.0], [67.0, 66.0],
+                                 [66.0, 65.0], [65.0, 64.0], [64.0, 63.0], [63.0, 62.0],
+                                 [62.0, 61.0], [61.0, 60.0]])
+
+        assert_array_equal(get_bounds_array(prefetch_dataset, dimension_path),
+                         expected_bounds_array)
+
+
+    def test_write_bounds(self):
+        """ Ensure that bounds data array is written to the dimension
+            dataset, both when the dimension variable is in the root group
+            and in a nested group.
+
+        """
+        varinfo_prefetch = VarInfoFromDmr('tests/data/ATL16_prefetch_group.dmr')
+        prefetch_dataset = Dataset('tests/data/ATL16_prefetch_group.nc4', 'r+')
+
+        # Expected variable contents in file.
+        expected_bounds_data = np.array([
+            [90.0, 89.0], [89.0, 88.0], [88.0, 87.0], [87.0, 86.0],
+            [86.0, 85.0], [85.0, 84.0], [84.0, 83.0], [83.0, 82.0],
+            [82.0, 81.0], [81.0, 80.0], [80.0, 79.0], [79.0, 78.0],
+            [78.0, 77.0], [77.0, 76.0], [76.0, 75.0], [75.0, 74.0],
+            [74.0, 73.0], [73.0, 72.0], [72.0, 71.0], [71.0, 70.0],
+            [70.0, 69.0], [69.0, 68.0], [68.0, 67.0], [67.0, 66.0],
+            [66.0, 65.0], [65.0, 64.0], [64.0, 63.0], [63.0, 62.0],
+            [62.0, 61.0], [61.0, 60.0]])
+
+        with self.subTest('Dimension variable is in the root group'):
```

Comment:
**An actual, non-nitpick request (gasp!)**

I was trying to think about what this test and the nested group test are trying to establish. I think it's a couple of things (and that they should all have assertions to cover them):

1) Did the bounds variable get created where you expected?
2) Does the bounds variable have the expected array? ✅ - tested already.
3) Does the bounds variable have the expected attributes? (Probably the least important assertion)
4) Does the dimension variable have the expected `bounds` metadata attribute?
5) Does the `VariableFromDmr` (part of the `varinfo_prefetch` object) have the `bounds` reference in it's `references` dictionary?

I think you're only explicitly asserting one of these things. The first thing (is the bounds variable where you expect) is sort of being tested, but not explicitly, and that would be a good thing to have as an outright assertion, before checking the data.

Reply:
Alrighty, so:

1. Did the bounds variable get created where you expected? ✅ - Nice (although maybe move it up above 2?)
2. Does the bounds variable have the expected array? ✅ - tested already.
3. Does the bounds variable have the expected attributes? (Probably the least important assertion) ❓ - Not sure I see this (but again, not super important)
4. Does the dimension variable have the expected bounds metadata attribute? ❓ - Not sure I see this (and just to state, this means the `bounds` attribute of the dimension variable in the NetCDF-4 file)
5. Does the VariableFromDmr (part of the varinfo_prefetch object) have the bounds reference in it's references dictionary? ✅ (although might be testing this twice)